### DISCUSSION TITLE
Arduino Due

### DISCUSSION THREAD

**Discussion by ososbob** [grblHAL/Controllers#6] [Q&A] [answered]
I have recently set up a 5 axis CNC using the grblHAL and an arduino due.  Almost everything works perfectly (although I have not set up limit switches yet)  I only have one small problem, for the life of me I cannot find the pin assignments for the probe.  I used the ramps 1.6 driver.  If anyone can help or point me in the right direction that would be great!.  I assumed it would be the normal grbl pin for arduino, the scl pin but that clearly does not work.  I was using the Protoneer  shield with an Uno prior to this build and my probe worked fine.  I have searched extensively on this but can't seem to find what I'm looking for.   
Thanks in advance.

**Comment by terjeio** [comment]
AFAIKT there is no pin officially designated as probe input on this board. However, there are plenty of pins available. Do you know if any of these are being used by other firmware for probe input and thus should be chosen by grblHAL as well? Or is it a matter of randomly choosing one and adding it to the board map?

**Comment by ososbob** [comment]
Thanks for the reply.  For this project I don't care what pin it is but looking at the maps if I understand it correctly it would appear pin 12 is available.  My only issue is I'm not sure how to compile the files for the Arduino Due to add a probe pin, I have been using your web builder and Bossa to load the board (works extremally well I might add).  I'm surprised this has not come up as having a probe pin assigned would make the Arduino Due just about plug and play for external stepper drivers.  I use the DMT542's and do not need any type of logic converter.  Really the only thing I added was a PWM to voltage module to drive my VFD.

**Reply by terjeio** [reply]
Ok, I have added pin 12 (D.08) - but not tested it yet. You may have to clear the browser cache to be able to build with the change.

**Comment by ososbob** [comment]
Terjeio,  I was out till late tonight but once I got home I saw your post and promptly uploaded the new file, attached my probe wire to 12 and ran a quick test.  I did have to invert the probe in openbuilds controller (I use this specifically as they advertise compatible with grblHal) As expected if I toudh pin 12 to ground the probe shows triggered :).  I will run some actual probing tomorrow but everything I seems to looks as if this worked.  If all testing goes well tomorrow I will post again and tag this as answered!  Thanks a ton!!!

**Comment by ososbob** [comment, marked as answer]
terjeio,
Well testing did not go as well as expected.  when actually probing the pin is triggered off and on for no apparent reason.  Further research shows grbl, using the Uno utilized  pin A5 (aka SCL ) as it has a internal pull up resistor and is a the two wire interface with A4.  The Arduino Due has an equal pins (Pins 20 and  21) as the primary two wire interface both with pull up resistors.  Since Due documents pin 21 as the "SCL" I think that is the pin which should be used as the probe for the Ramps 1.6 map.  If you can could we change the probe from 12 to 21 and I can test it to make sure that works.  I'm sure you are busy with other stuff but I really do appreciate your efforts and I also believe this  benefit others.

Bob

Update:  I have figured out how to edit the given driver files and change the pins on my own.  However, I need some direction in understanding the due pins to "PIO" you have in the code  for example you have 
#define PROBE_PORT          PIOD
#define PROBE_PIN           8  
for Due pin 12
 .  If there is some documentation that can point me in the right direction so I can determine if the port needs to be an "A,B,C or D" and how the pin numbering relates to the actual pinout of the Due that would be great.

Update 2:  I figured out the relationships and how to program the due via IDE and Bossa, so changing a pin is no problem for me now so I'm going to mark this discussion as answered.  However, the probing is still not working as desired as I still get phantom triggers when moving or probing.  Additionally, probing is moving in the opposite direction which may be the openbuilds controller issue.  I'm sure I can write a macro to correct that.  If anyone knows where I can get help or the correct discussion to explore why phantom triggers are happening, I would really appreciate the pointing in the right direction.

**Comment by ososbob** [comment]
Final update.  All issue have been resolved!  Once I made sure the ground was connected to the spindle bit and the Pin (in this case 21 or SCL) was connected to the touch plate and a filter (0.01 uf Cap and 100 ohm resistor) was in place all worked as expected.  Give someone enough time and they can fix their own problems.  :)
